data-prepper-plugins/aggregate-processor/src/main/java/org/opensearch/dataprepper/plugins/processor/aggregate/actions/TailSamplerAggregateAction.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-Limense-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.plugins.processor.aggregate.actions;
+
+import org.opensearch.dataprepper.model.annotations.DataPrepperPlugin;
+import org.opensearch.dataprepper.model.annotations.DataPrepperPluginConstructor;
+import org.opensearch.dataprepper.model.event.Event;
+import org.opensearch.dataprepper.plugins.processor.aggregate.AggregateAction;
+import org.opensearch.dataprepper.plugins.processor.aggregate.AggregateActionInput;
+import org.opensearch.dataprepper.plugins.processor.aggregate.AggregateActionResponse;
+import org.opensearch.dataprepper.plugins.processor.aggregate.GroupState;
+import org.opensearch.dataprepper.expression.ExpressionEvaluator;
+
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Random;
+import java.time.Duration;
+import java.time.Instant;
+
+/**
+ * An AggregateAction that combines multiple Events into a single Event. This action 
+ * 
+ * @since 2.1
+ */
+@DataPrepperPlugin(name = "tail_sampler", pluginType = AggregateAction.class, pluginConfigurationType = TailSamplerAggregateActionConfig.class)
+public class TailSamplerAggregateAction implements AggregateAction {
+    static final String LAST_RECEIVED_TIME_KEY = "last_received_time";
+    static final String EVENTS_KEY = "events";
+    static final String ERROR_STATUS_KEY = "error_status";
+    private final double percent;
+    private final Duration waitPeriod;
+    private final ExpressionEvaluator<Boolean> expressionEvaluator;
+    private final String errorCondition;
+    private boolean shouldCarryGroupState;
+
+    @DataPrepperPluginConstructor
+    public TailSamplerAggregateAction(final TailSamplerAggregateActionConfig tailSamplerAggregateActionConfig, final ExpressionEvaluator<Boolean> expressionEvaluator) {
+        percent = tailSamplerAggregateActionConfig.getPercent();
+        waitPeriod = tailSamplerAggregateActionConfig.getWaitPeriod();
+        errorCondition = tailSamplerAggregateActionConfig.getErrorCondition();
+        this.expressionEvaluator = expressionEvaluator;
+        shouldCarryGroupState = true;
+    }
+
+    @Override
+    public boolean shouldCarryState() {
+        return shouldCarryGroupState;
```

Comment:
This could result in thread issues.

If you have multiple groups at the same time, then the value of this will change between calls to `concludeGroup`. So this value may not match the expected value.

I think a better approach would be to change the return type of `concludeGroup` to a new class - `AggregationActionOutput`.

```
public class AggregationActionOutput {
  private List<Event> events;
  private boolean shouldCarryState;
}
```

I'm not quite sure we do need to carry this state though. See my other comment.